src/ocrd/processor/base.py

```diff
@@ -188,20 +233,137 @@ def verify(self):
+                            continue
+                        try:
+                            input_files[i] = self.workspace.download_file(input_file)
+                        except ValueError as e:
+                            log.error(repr(e))
+                            log.warning("skipping file %s for page %s", input_file, input_file.pageId)
+                    self.process_page_file(*input_files)
+            except NotImplementedError:
+                # fall back to deprecated method
+                self.process()
 
-    def add_metadata(self, pcgts):
+    def process_page_file(self, *input_files) -> None:
+        """
+        Process the given ``input_files`` of the :py:attr:`workspace`,
+        representing one physical page (passed as one opened
+        :py:class:`~ocrd_models.OcrdFile` per input fileGrp)
+        under the given :py:attr:`parameter`, and make sure the
+        results get added accordingly.
+
+        (This uses process_page_pcgts, but can be overridden by subclasses
+        to handle cases like multiple fileGrps, non-PAGE input etc.)
+        """
+        log = getLogger('ocrd.processor.base')
+        input_pcgts = [None] * len(input_files)
+        page_id = input_files[0].pageId
+        for i, input_file in enumerate(input_files):
+            # FIXME: what about non-PAGE input like image or JSON ???
+            log.debug("parsing file %s for page %s", input_file.ID, input_file.pageId)
+            try:
+                input_pcgts[i] = page_from_file(input_file)
+            except ValueError as e:
+                log.info("non-PAGE input for page %s: %s", page_id, e)
+        output_file_id = make_file_id(input_files[0], self.output_file_grp)
+        output_pcgts = self.process_page_pcgts(*input_pcgts, output_file_id=output_file_id, page_id=page_id)
+        if isinstance(output_pcgts, (list, tuple)):
+            output_images = output_pcgts[1:]
+            output_pcgts = output_pcgts[0]
+            for output_image_pil, output_image_id, output_image_path in output_images:
+                self.workspace.save_image_file(
+                    output_image_pil,
+                    output_image_id,
+                    self.output_file_grp,
+                    page_id=page_id,
+                    file_path=output_image_path)
+        output_pcgts.set_pcGtsId(output_file_id)
+        self.add_metadata(output_pcgts)
+        # FIXME: what about non-PAGE output like JSON ???
+        self.workspace.add_file(file_id=output_file_id,
+                                file_grp=self.output_file_grp,
+                                page_id=page_id,
+                                local_filename=os.path.join(self.output_file_grp, output_file_id + '.xml'),
+                                mimetype=MIMETYPE_PAGE,
+                                content=to_xml(output_pcgts))
+
+    def process_page_pcgts(self, *input_pcgts, output_file_id : str = None, page_id : str = None) -> OcrdPage:
```

Comment:
```suggestion
    def process_page_pcgts(self, *input_pcgts, output_file_id : Optional[str] = None, page_id : Optional[str] = None) -> OcrdPage:
```

Reply:
Are you sure that `*input_pcgts : OcrdPage` does mean "any number of `OcrdPage` objects?

Reply:
Potentially that is `input_pcgts: List[OcrdPage]`?

Reply:
By [this SO answer](https://stackoverflow.com/a/48909674) @kba is correct while `List[OcrdPage]` would mean "any number of list of OcrdPage objects".